File: server/util/overpass.py
from aiohttp import ClientSession
from urllib.parse import quote_plus


OVERPASS_API_INTERPRETER = "https://maps.mail.ru/osm/tools/overpass/api/interpreter"


async def query_osm(query: str, bbox: tuple[int, int, int, int] | None = None) -> list:
    if "{{bbox}}" in query:
        if not bbox:
            raise ValueError("{{bbox}} in query, but it wasn't provided")
        query = query.replace("{{bbox}}", ", ".join(map(str, bbox)))

    async with ClientSession() as session:
        async with session.post(
            OVERPASS_API_INTERPRETER, data="data=" + quote_plus("[out:json];" + query)
        ) as response:
            return (await response.json())["elements"]


async def find_path(start: tuple[int, int], end: tuple[int, int]):
    await query_osm(
        f"""
        [out:json];
        way
            [railway=rail]
            [usage=main]
            ({{bbox}});
        out geom;
        """,
        (*start, *end),
    )

File: client/src/geo/overpass.ts
const OVERPASS_API_INTERPRETER = 'https://maps.mail.ru/osm/tools/overpass/api/interpreter';

export const osmQuery = async (ql: string, bbox?: [number, number, number, number]) => {
  if (ql.includes('{{bbox}}')) {
    if (bbox === undefined)
      throw new TypeError("{{bbox}} was found in query, but it wasn't provided");
    ql = ql.replace('{{bbox}}', bbox.join(', '));
  }

  return (
    await (
      await fetch(OVERPASS_API_INTERPRETER, {
        method: 'POST',
        body: 'data=' + encodeURIComponent('[out:json];' + ql),
      })
    ).json()
  ).elements;
};

export type Way = {
  type: 'way';
  id: number;
  nodes: number[];
};

export type Node = {
  type: 'node';
  id: number;
  lat: number;
  lon: number;
};

export const queryWay = async (
  start: [number, number],
  end: [number, number],
): Promise<(Way | Node)[]> => {
  return osmQuery(`
  way
    [railway=rail]
    [usage=main]
    (${[...start, ...end].join(', ')});
  (._;>;);
  out body;
  `);
};
